### ISSUE TITLE
Client errors writing to Oak in AEM 6.1

### ISSUE THREAD

**Issue by jdigger** [TWCable/grabbit#70]
Periodically this shows up in the logs, failing the job. Seems to be related to https://issues.apache.org/jira/browse/OAK-2527 but that fix should be in AEM 6.1.

```
02.09.2015 21:52:14.141 *ERROR* [clientJobLauncherTaskExecutor-5] com.twcable.grabbit.client.batch.steps.jcrnodes.JcrNodesWriter Exception writing JCR Nodes to current JCR Session : session-259564.
org.springframework.batch.core.listener.StepListenerFailedException: Error in afterWrite.
        at org.springframework.batch.core.listener.MulticasterBatchListener.afterWrite(MulticasterBatchListener.java:257)
        at org.springframework.batch.core.step.item.SimpleChunkProcessor.doAfterWrite(SimpleChunkProcessor.java:167)
        at org.springframework.batch.core.step.item.SimpleChunkProcessor.doWrite(SimpleChunkProcessor.java:152)
        at org.springframework.batch.core.step.item.SimpleChunkProcessor.write(SimpleChunkProcessor.java:274)
        at org.springframework.batch.core.step.item.SimpleChunkProcessor.process(SimpleChunkProcessor.java:199)
        at org.springframework.batch.core.step.item.ChunkOrientedTasklet.execute(ChunkOrientedTasklet.java:75)
        at org.springframework.batch.core.step.tasklet.TaskletStep$ChunkTransactionCallback.doInTransaction(TaskletStep.java:395)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:131)
        at org.springframework.batch.core.step.tasklet.TaskletStep$2.doInChunkContext(TaskletStep.java:267)
        at org.springframework.batch.core.scope.context.StepContextRepeatCallback.doInIteration(StepContextRepeatCallback.java:77)
        at org.springframework.batch.repeat.support.RepeatTemplate.getNextResult(RepeatTemplate.java:368)
        at org.springframework.batch.repeat.support.RepeatTemplate.executeInternal(RepeatTemplate.java:215)
        at org.springframework.batch.repeat.support.RepeatTemplate.iterate(RepeatTemplate.java:144)
        at org.springframework.batch.core.step.tasklet.TaskletStep.doExecute(TaskletStep.java:253)
        at org.springframework.batch.core.step.AbstractStep.execute(AbstractStep.java:195)
        at org.springframework.batch.core.job.SimpleStepHandler.handleStep(SimpleStepHandler.java:141)
        at org.springframework.batch.core.job.flow.JobFlowExecutor.executeStep(JobFlowExecutor.java:64)
        at org.springframework.batch.core.job.flow.support.state.StepState.handle(StepState.java:60)
        at org.springframework.batch.core.job.flow.support.SimpleFlow.resume(SimpleFlow.java:151)
        at org.springframework.batch.core.job.flow.support.SimpleFlow.start(SimpleFlow.java:130)
        at org.springframework.batch.core.job.flow.FlowJob.doExecute(FlowJob.java:135)
        at org.springframework.batch.core.job.AbstractJob.execute(AbstractJob.java:301)
        at org.springframework.batch.core.launch.support.SimpleJobLauncher$1.run(SimpleJobLauncher.java:134)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: null
        at com.google.common.base.Preconditions.checkState(Preconditions.java:134)
        at org.apache.jackrabbit.oak.plugins.segment.Segment.<init>(Segment.java:147)
        at org.apache.jackrabbit.oak.plugins.segment.Segment.<init>(Segment.java:138)
        at org.apache.jackrabbit.oak.plugins.segment.file.FileStore.readSegment(FileStore.java:833)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentTracker.getSegment(SegmentTracker.java:128)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentId.getSegment(SegmentId.java:108)
        at org.apache.jackrabbit.oak.plugins.segment.Record.getSegment(Record.java:82)
        at org.apache.jackrabbit.oak.plugins.segment.MapRecord.size(MapRecord.java:142)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentWriter.writeMapBucket(SegmentWriter.java:513)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentWriter.writeMapBucket(SegmentWriter.java:511)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentWriter.writeMapBucket(SegmentWriter.java:511)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentWriter.writeMap(SegmentWriter.java:711)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentWriter.writeNode(SegmentWriter.java:1099)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentWriter.writeNode(SegmentWriter.java:1101)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentNodeBuilder.getNodeState(SegmentNodeBuilder.java:100)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentNodeBuilder.getNodeState(SegmentNodeBuilder.java:34)
        at org.apache.jackrabbit.oak.plugins.index.property.PropertyIndexEditor.leave(PropertyIndexEditor.java:282)
        at org.apache.jackrabbit.oak.plugins.index.IndexUpdate.leave(IndexUpdate.java:221)
        at org.apache.jackrabbit.oak.spi.commit.VisibleEditor.leave(VisibleEditor.java:63)
        at org.apache.jackrabbit.oak.spi.commit.CompositeEditor.leave(CompositeEditor.java:74)
        at org.apache.jackrabbit.oak.spi.commit.EditorDiff.process(EditorDiff.java:56)
        at org.apache.jackrabbit.oak.spi.commit.EditorHook.processCommit(EditorHook.java:54)
        at org.apache.jackrabbit.oak.spi.commit.CompositeHook.processCommit(CompositeHook.java:61)
        at org.apache.jackrabbit.oak.spi.commit.CompositeHook.processCommit(CompositeHook.java:61)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentNodeStore$Commit.prepare(SegmentNodeStore.java:405)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentNodeStore$Commit.optimisticMerge(SegmentNodeStore.java:428)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentNodeStore$Commit.execute(SegmentNodeStore.java:484)
        at org.apache.jackrabbit.oak.plugins.segment.SegmentNodeStore.merge(SegmentNodeStore.java:162)
        at org.apache.jackrabbit.oak.spi.state.ProxyNodeStore.merge(ProxyNodeStore.java:43)
        at org.apache.jackrabbit.oak.core.MutableRoot.commit(MutableRoot.java:247)
        at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.commit(SessionDelegate.java:313)
        at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.save(SessionDelegate.java:459)
        at org.apache.jackrabbit.oak.jcr.session.SessionImpl$8.perform(SessionImpl.java:435)
        at org.apache.jackrabbit.oak.jcr.session.SessionImpl$8.perform(SessionImpl.java:432)
        at org.apache.jackrabbit.oak.jcr.delegate.SessionDelegate.perform(SessionDelegate.java:216)
        at org.apache.jackrabbit.oak.jcr.session.SessionImpl.perform(SessionImpl.java:140)
        at org.apache.jackrabbit.oak.jcr.session.SessionImpl.save(SessionImpl.java:432)
        at sun.reflect.GeneratedMethodAccessor36.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:483)
        at org.apache.sling.jcr.base.SessionProxyHandler$SessionProxyInvocationHandler.invoke(SessionProxyHandler.java:113)
        at com.sun.proxy.$Proxy8.save(Unknown Source)
        at com.twcable.grabbit.client.batch.steps.jcrnodes.JcrNodesWriter.afterWrite(JcrNodesWriter.groovy:50)
        at org.springframework.batch.core.listener.CompositeItemWriteListener.afterWrite(CompositeItemWriteListener.java:60)
        at org.springframework.batch.core.listener.MulticasterBatchListener.afterWrite(MulticasterBatchListener.java:254)
        ... 25 common frames omitted
```

The condition that's failing is the `checkState` below:
```java
public Segment(SegmentTracker tracker, SegmentId id, ByteBuffer data, SegmentVersion version) {
    this.tracker = checkNotNull(tracker);
    this.id = checkNotNull(id);
    this.data = checkNotNull(data);
    if (id.isDataSegmentId()) {
        byte segmentVersion = data.get(3);
        checkState(data.get(0) == '0'
                && data.get(1) == 'a'
                && data.get(2) == 'K'
                && SegmentVersion.isValid(segmentVersion));
        this.refids = new SegmentId[getRefCount()];
        this.refids[0] = id;
        this.version = SegmentVersion.fromByte(segmentVersion);
    } else {
        this.refids = null;
        this.version = version;
    }
}
```

**Comment by jdigger** [comment]
Maybe this is the 6.1 equivalent to GH-1 or GH-4? My best (shot-in-the-dark) guess is that it's being caused by concurrency, so maybe it's related to GH-38?

**Comment by jdigger** [comment]
Apparently this destroys the JCR and Grabbit will "never run again" on the client, per @hsfear

**Comment by hsfear** [comment]
Maybe a bit of hyperbole.  It doesn't "destroy" the jcr.  And you can run grabbit.  It just doesn't seem to succeed on subsequent runs.  And the exception is thrown continuously every 5 seconds or so.  Which definitely fills up the logs.

**Comment by sagarsane** [comment]
Since we now have this commit in : https://github.com/TWCable/grabbit/commit/b82662cfa1cdb2b2e74089a351824ac9f03d7aba; could try to enable debug logging to see if there is a specific node / kind of node where this exception occurs.

**Comment by peroks** [comment]
We get this error on the client side when the server throws "ava.lang.OutOfMemoryError: Requested array size exceeds VM limit", see "OutOfMemoryError (VM limit exceeded)" https://github.com/TWCable/grabbit/issues/76

**Comment by jbornemann** [comment]
I wonder if this is still an issue. I haven't seen something like this in some time

**Comment by jdigger** [comment]
Have not seen or heard of this in a long time. Suspect that the Oak fix has finally shown up in one of the AEM 6.1 hotfixes.
